examples/splinterdb_custom_ipv4_addr_sortcmp_example.c

```diff
@@ -411,10 +346,14 @@ do_iterate_all(splinterdb *spl_handle, int num_keys)
 static int
 do_iterate_from(splinterdb *spl_handle, const char *from_key)
 {
-   ex_msg("Iterate through all the keys starting from '%s':\n", from_key);
+   printf("Iterate through all the keys starting from '%s':\n", from_key);
 
    splinterdb_iterator *it = NULL;
 
+   // Use min-key if user has not specified any start-from-key
+   if (!from_key) {
+      from_key = IP4ADDR_MIN_KEY;
```

Comment:
@rtjohnso - We should review this carefully.

I discovered that when user-specified sort-key comparison is used, specifying `NULL_SLICE` as initial start_key for "iterate_all" does not work. Something about the `iterator_init()` isn't getting setup correctly. And scan terminates w/o returning any kv-pair.

I reasoned that if user is using custom-sorting, just as they specified `min_key` while configuring SplinterDB, they will need to also specify this `min_key` when doing iterate-for-all w/o specifying start_key.

Makes sense?

Reply:
This seems like a bug to me

Reply:
Hmmm ... @ajhconway - I did debug this somewhat, to see what was going wrong in `iterator_init()`.

Honestly, I did not really get to the bottom of it, and "convinced" myself that specifying `from_key` as shown in the code is the acceptable fix.

@rtjohnso -- What's your call? 

I'd like to try to merge this PR in as-is, register a potential issue and debug that off-line. If we do conclude that it's a bug, I'd like to fix the code and the example code in a follow-on. That way, this PR's progress is not blocked by a potential bug.

If you both think that this really a bug, I will go back and investigate this via a unit-test, and debug it fully.
Let me get back to you on that.

Reply:
@ajhconway , @rtjohnso -- We should discuss this in our 1pm tomorrow, but for the record, here's my assessment based on debugging this issue.

----
I found what was a piece of "compensating code" in this PR that was creating / masking some issue which Alex pointed out as "the bug". I tracked it down to this piece of code in this sample program itself, in the custom-comparator function; i.e. here:

```
242 ip4_ipaddr_keycmp(const char  *key1,
243                   const size_t key1_len,
244                   const char  *key2,
245                   const size_t key2_len)
246 {
247    // Deal with the case of initial keys, where one of the pair maybe uninit'ed
248    if ((key1_len == 0) || (key2_len == 0)) {
249       return KEYCMP_RV_KEY1_EQ_KEY2;
250    }
251
```

----
On L249, we return "keys-are-equal", leading to some unexpected behaviour. 

When I took this block out, this example program runs cleanly even for the `do_iterate_from()` case when `start_key = NULL`.

----
I then wondered and debugged this code in `splinterdb_iterator_init()`:

```
800 splinterdb_iterator_init(const splinterdb     *kvs,      // IN
[...]

814    char start_key_buffer[MAX_KEY_SIZE] = {0};
815    if (!slice_is_null(start_key)) {
816       int rc = encode_key(MAX_KEY_SIZE, start_key_buffer, start_key);
817       if (rc != 0) {
818          return rc;
819       }
820    }
821
822    platform_status rc = trunk_range_iterator_init(
823       kvs->spl, range_itor, start_key_buffer, NULL, UINT64_MAX);
```

On L 823, even when `slice_is_null(start_key()` is TRUE, we are passing-down `start_key_buffer`, which would be uninitialized.

I played around with passing-down NULL to the callee here when `start_key` is NULL  [which seems to be the reasonable thing to do ]  ... and that seems to work too.

So ... now all this is very confusing to me.

It's not clear what the right interface should be. I don't want to hack-patch this ... so let's talk tomorrow.